README.md
## IOS混淆验证服务器端

### 原理：模拟源服务器验证程序返回数据结构，替换clang中的接口地址达到验证通过的目的


CHANGELOG
2017/12/25 完成校验接口，替换原有clang中的服务器验证地址

common/co.js
/**
 * Created by smallwolf on 2017/8/4.
 */
'use strict';

let myHandles = {};
let createDebug = require('debug');
let debug = createDebug('mjserver:resp');

function sendFormat(res, data, templ) {

    res.format({
        json: () => {
            res.json(data);
        },
        html: () => {
            res.send(data);
        }
    });

}

function executor(fn) {
    myHandles[fn.name] = fn;
    return function runHandle(req, res, next) {
        fn(req, res, next)
            .then(data => {
                if (data === undefined || req.timedout) {
                    return;
                }
                if (data === SM.constants.NEXT_SYMBOL) {
                    return next();
                }
                let ret = {
                    errcode: 0,
                    errmsg: 'OK',
                };
                debug(data);
                ret = Object.assign({}, ret, data);
                sendFormat(res, ret);
            })
            .catch(err => {
                console.error(`${req.method} ${req.url}`);
                console.error(err.stack || err);
                if (true) {
                    delete err.stack;
                }
                if (!err.errcode) {
                    err = new SM.Exception(SM.errors.ERROR_INTERNAL);
                }
                sendFormat(res, err, 'error');
            });
    };
}


module.exports = executor;

common/app.js
let globalNS = require('./define_prop_global.js');

var express = require('express');
var ejs = require('ejs');
var path = require('path');
var favicon = require('serve-favicon');
var logger = require('morgan');
var cookieParser = require('cookie-parser');
var bodyParser = require('body-parser');
var timeout = require('connect-timeout');
var session = require('express-session');


var router = require('../routes/index');

var app = express();

app.use(timeout(100 * 1000, {respond: false}));
app.use(logger('dev'));
app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));
app.use(cookieParser());
app.use(session({
    key: "clang-sid",
    secret: 'HdVWOPn2wpTrmWahSuS5ftVgwUtEfBkJ',
    resave: true,
    saveUninitialized: true,
    cookie: {secure: false, maxAge: 30 * 24 * 3600 * 1000, path: '/', httpOnly: false, /* domain: '*'*/}
}));
app.use(async (req, res, next) => {
    if (!req.session.user) {
        req.session.user = {};
    }

    req.user = req.session.user;
    app.locals.req = req;
    next();
});

app.use('/clang', router);


// catch 404 and forward to error handler
app.use(SM.co(async (req, res, next) => {
    res.status(404);
    throw new SM.Exception(SM.errors.ERROR_NOT_PAGE);
}));

module.exports = app;


routes/index.js
var express = require('express');
var router = express.Router();
var debug = require('debug')('clangcheck:clang:check');

// sign = mac + ip + clang_id + bundleId + '!@#$%'
// http://47.104.80.5:5876/api/handlers/check?mac=dc:a9:04:86:c6:ff&ip=192.168.10.77&id=826232c5faf849a3a05aa2a2f7ca0952&bundleId=com.bjzhangyun.sxmj&sign=b80212b0e2159f72bb17ec9ca5e72e22

// clangswift no|2781044552|1d67e201540142de3ff050a9a83ea2a8            result|~Timestamp(s)>>>0|sign
// clangswift sign = md5(result + Timestamp(s) + '!@#$%')

// clanginfo 13|2781039533|72f7b253f49a4f62d7a6fcc6441eb1b3   count|~Timestamp(s)>>>0|sign
// clanginfo sign = md5(count + Timestamp(s) + '!@#$%')

router.get('/check', function (req, res) {
    let data = req.query;
    data = Object.assign({mac: null, ip: null, id: null, bundleId: null, sign: null}, data);

    // debug("MAC:%s, IP:%s, userId:%s, bundleId:%s | md5(mac + ip + id + bundleId + priStr) = %s", data.mac, data.ip, data.id, data.bundleId, data.sign);
    let reqRealSign = SM._.md5(`${data.mac}${data.ip}${data.id}${data.bundleId}!@#$%`);
    if (reqRealSign !== data.sign) {
        res.send('failed');
        return;
    }

    // 时间戳取反的无符号整数
    let curTimestamp = Math.floor(Date.now() / 1000);

    let notTimestamp = (~curTimestamp) >>> 0;

    let ret = 'yes';

    let sign = SM._.md5(`${ret}${curTimestamp}!@#$%`);
    ret = `${ret}|${notTimestamp}|${sign}`;
    debug(ret);
    res.send(ret);
});

module.exports = router;


common/errors.js
/**
 * Created by smallwolf on 2017/8/4.
 */

module.exports = {
  ERROR_INTERNAL: [10000, '服务器内部错误'],
  ERROR_REQUEST: [1000, '请求出错'],
  ERROR_SIGN_CHECK: [1001, 'sign校验失败'],
  ERROR_PARAM_CHECK: [1002, '参数不正确'],
  ERROR_NOT_PAGE: [1003, '页面不存在'],
  ERROR_NO_LOGIN: [1004, '用户未登录'],
  ERROR_NO_ROOM_OWNER: [1006, '不是房主无法解散'],
  ERROR_DISSOLVE_ROOM_FAIL: [1007, '解散房间失败'],
  ERROR_REQUEST_TIMEOUT: [1008, '请求超时'],
  ERROR_INVITED: [1009, '已经绑定代理'],
  ERROR_TOKEN_CHECK: [1010, '无权限操作'],
  ERROR_USER_EXIST: [1011, '用户已存在'],
  ERROR_GEM_NOT_ENUGHT: [1012, '房卡数量不足'],
  ERROR_USER_IN_ROOM: [1013, '玩家已经在房间中'],
  ERROR_USER_CREATE_ROOM: [1014, '创建房间失败'],
  ERROR_USER_ENTER_ROOM: [1015, '进入房间失败'],
  ERROR_NOT_FOUND_GAME_SERVER: [1016, '服务器已满'],
  ERROR_USED_ORDER: [1017, '该订单已经充值过了'],
  ERROR_USED_ACTIVITY_CHANCE: [1018, '已经分享过了'],
  ERROR_EXISTED_IN_ROOM: [1019, '你已经在房间中'],
  ERROR_ALREADY_GOT: [1020, '你已经领取过了'],
  ERROR_NOT_ABILITY: [1021, '你还有达到领取资格'],
  ERROR_DEALING_DISSOLVE:[1022, '系统已经正在进行结算了'],
  ERROR_ORDER_USER:[1023, '订单信息不存在'],
  ERROR_WECHAT_LOGIN:[1024, '微信登陆失败'],
  ERROR_NOT_PROXY:[1025, '该用户不是代理'],
  ERROR_NOT_USER:[1026, '该用户不存在'],
  ERROR_REGEISTER_USER:[1027, '注册失败'],
  ERROR_EXIST_USER:[1028, '用户已存在'],

  ERROR_DATA_NOEXISTED: [2000, '数据不存在'],
  ERROR_DATA_UPDATE_FAILED: [2001, '数据修改失败'],
  ERROR_DATA_CREATE_FAILED: [2002, '创建失败'],
  ERROR_GEN_TICKET_ERROR: [2003, '生成微信JS TICKET失败'],
  
  ERROR_REQ_THIRD_PARAMS: [3000, '参数错误'],
  ERROR_REQ_KEY: [3001, '未找到对应的配置信息'],
  ERROR_RECEIPT_NULL: [3002, '未找到票据信息'],
  ERROR_RECEIPT_PRODUCT_ID: [3003, '购买的商品不存在']
};

common/constants.js
/**
 * Created by Administrator on 2017/7/31 0031.
 */
let _constant = {};
let CONSTANT = new Proxy(_constant, {
    set: function (target, name, value, receiver) {
        if (_constant[name]) {
            throw new Error(`${__filename} has some dup key. key:${name}`);
        }
        let success = Reflect.set(target, name, value, receiver);
        return success;
    }
});
CONSTANT.SESSION_KEY = 'test';
CONSTANT.NEXT_SYMBOL = Symbol['NEXT_SYMBOL'];
CONSTANT.MJ_TYPE = {
    TONG: 0,
    TIAO: 1,
    WAN: 2,
    FENG: 3,
    ZHONG: 4,
    FA: 5,
    BAIBAN: 6,

};
CONSTANT.MANUAL_TYPE = {
    GET_ROOMS: 'rooms',
    GET_ROOM_DETAIL: 'room',
    DEAL_INIT: 'initial',
    DEAL_MANUAL: 'manual',
    SET_REMAIN: 'remain',
    DEAL_SPEC_BY_USERID: 'userspec',
};
CONSTANT.STATISTIC_INTERVAL = 5 * 60 * 1000;
CONSTANT.TABLE_PREFIX = 't_';
CONSTANT.DEFAULT_COIN = 0;
CONSTANT.DEFAULT_GEM = 10;
CONSTANT.DEFAUTLT_INVITION_GEM = 3;
CONSTANT.DEFAUTLT_SHARE_GEM = 3;
CONSTANT.OAUTH_INFO = {
    getCode: {
        url: (appId, redirectUri, scope, state) => {
            return `https://open.weixin.qq.com/connect/oauth2/authorize?appid=${appId}&redirect_uri=${redirectUri}&response_type=code&scope=${scope}&state=${state}#wechat_redirect`;
        }
    },
    getAccessToken: {
        url: "https://api.weixin.qq.com/sns/oauth2/access_token"
    },
    getUserInfo: {
        url: "https://api.weixin.qq.com/sns/userinfo"
    },
    getAccessToken2: {
        url: 'https://api.weixin.qq.com/cgi-bin/token'
    },
    getTicket: {
        //access_token=ACCESS_TOKEN&type=jsapi
        url: 'https://api.weixin.qq.com/cgi-bin/ticket/getticket'
    },
    getFile: {
        //access_token=ACCESS_TOKEN&media_id=jsapi
        url: 'http://file.api.weixin.qq.com/cgi-bin/media/get'
    }
};
CONSTANT.RESULT_TYPE = {
    SUCCESS: 1,
    FAIL: 0
};
CONSTANT.GEM_COST_TYPE = {
    RECHAGE: 10,
    CREATE_ROOM_COST: 1,
    CREATE_ROOM_NOJOIN_COST: 2,
    CREATE_ROOM_RETURNED: 6,
    SYS_COST: 5,
    SYS_SEND: 4,
    PROXY_ASSIGN: 3,//代理充值
    PROXY_RECHARGE_COST: 7,//代理充值消耗
};
CONSTANT.ROOM_TYPE = {
    NORMAL: 1,
    WAIT_ROOM: 2
};
CONSTANT.PAY_STATUS = {
    SUBMIT_ORDER: 1,
    WAIT_PAYING: 2,
    CANCLE_PAY: 3,
    SUCCESS_PAY: 4,
    FAIL_PAY: 5
};
CONSTANT.INVITION_STATUS = {
    NO_ACTIVE: 1,
    ACTIVE: 2
};

CONSTANT.COMMON_STATUS = {
    NORMAL: 1,
    LOCK: 2,
    DELETE: -1,
};

CONSTANT.GENDER = {
    SECRET: 0,
    MAN: 1,
    WOMAN: 2
};

CONSTANT.PRODUCTS_MAP = {
    "1": {price: 0.01, gems: 1, name: '1个房卡'},
    "2": {price: 0.01, gems: 5, name: '5个房卡'},
    "3": {price: 0.01, gems: 10, name: '10个房卡'},
    "4": {price: 0.01, gems: 20, name: '20张房卡'},
    "5": {price: 0.01, gems: 40, name: '40个房卡'},
    "6": {price: 0.01, gems: 80, name: '80个房卡'},
    "7": {price: 0.01, gems: 160, name: '160个房卡'},
};
CONSTANT.PAY_TYPE = {
    // 2	汇元个人会员余额支付
    // 10	骏卡支付（话费通卡）
// 12	手机卡支付
// 13	神州行
// 14	联通卡
// 15	电信卡
// 18	银行卡快捷支付
// 22	支付宝
// 30	微信支付
// 35	盛大卡支付
// 40	第三方卡支付
// 41	盛大一卡通
// 42	网易一卡通
// 44	完美一卡通
};
CONSTANT.GUI_COUNT = [0, 1, 2, 3, 4];
CONSTANT.WAN_FA_MAP = {
    cg: 136,
    fhz: 136,
    bdf: 108,
    kpqz: 52,
    tbnn: 52,
    tbnn1: 52,
};
CONSTANT.ACCESS_TOKEN_CACHE_KEY = 'wechat:access_token';
CONSTANT.JS_TICKET_CACHE_KEY = 'wechat:js_ticket';

CONSTANT.PROXY_LEVEL = {
    ZONGDAI: 1,
    HUIZHANG: 2,
    GAOJIDAILI: 3,
    DAILI: 4
};
CONSTANT.ROOM_PAY_TYPE = {
    AA: 1,
    SINGLE: 2
};

module.exports = CONSTANT;

common/exception.js
/**
 * Created by smallwolf on 2017/8/4.
 */
'use strict';

let _stackTraceLimit = 10;

class Exception extends Error {

    static get stackTraceLimit() {
        return _stackTraceLimit;
    };

    static set stackTraceLimit(value) {
        _stackTraceLimit = value;
    };

    static captureStackTrace(targetObject, constructorOpt) {
        Error.captureStackTrace(targetObject, constructorOpt);
    }

    constructor(code, msg, errData) {
        // Error.captureStackTrace = (targetObject, constructorOpt) => {
        //   Error.captureStackTrace(targetObject, constructorOpt);
        // };
        let tempMsg = '';
        if (SM._.isArray(code)) {
            tempMsg = code[1];
            code = code[0];
        }
        if (SM._.isFunction(tempMsg)) {
            msg = tempMsg(msg);
        }
        else if (SM._.isEmpty(msg)) {
            msg = tempMsg;
        }
        Error.stackTraceLimit = _stackTraceLimit;
        super(`${Exception.name} -> ${msg}[${code}]`);
        this.errcode = code;
        this.errmsg = msg || 'Unknown error';
        this.fallback = errData;
    }
}

Exception.stackTraceLimit = 5;

module.exports = Exception;

common/define_prop_global.js
/**
 * Created by smallwolf on 2017/8/4.
 */
'use strict';
let debug = require('debug')('mjserver:global:attach');
let publicProps = new Proxy({}, {
    set: function (target, name, value, receiver) {
        if (publicProps[name]) {
            throw new Error(`${__filename} has some dup key. key:${name}`);
        }
        let success = Reflect.set(target, name, value, receiver);
        debug('attach prop: %s to global', name);
        return success;
    }
});
global.SM = publicProps;

publicProps.moment = require('moment');
publicProps.errors = require('./errors.js');
publicProps.Exception = require('./exception.js');
publicProps.constants = require('./constants.js');
publicProps.co = require('./co.js');
publicProps._ = require('lodash');
require('./lodash_extend.js')(publicProps._);

module.exports = publicProps;

common/lodash_extend.js
/**
 * Created by smallwolf on 2017/8/4.
 */
"use strict";
let request = require('request');
let crypto = require('crypto');
let iconv = require('iconv-lite');

function extendsFunc(ns) {
  ns.Request = async (opts) => {
    opts.timeout = opts.timeout || 30 * 1000;
    opts.json = opts.json === undefined ? true : opts.json;
    let throwError = opts.doesNotThrow === undefined ? true : !opts.doesNotThrow;
    delete opts.doesNotThrow;
    return await new Promise((resolve, reject) => {
      request(opts,
        function (error, response, body) {
          if (error && throwError) {
            return reject(new MJSERVER.Exception(MJSERVER.errors.ERROR_REQUEST, error.message));
          }
          else if (error && !throwError) {
            console.error('request error:', error);
            resolve(0);
          }
          if (opts.json) {
            if (body.errcode === 0) {
              resolve(body);
            }
            else if (throwError) {
              reject(new MJSERVER.Exception(body.errcode, `[${opts.url}]` + body.errmsg || body));
            }
            else {
              console.error('request error:', `[${opts.url}]` + body);
              resolve(0);
            }
          }
          else {
            resolve(body);
          }
        });
    });
  };
  ns.Request.post = async (url, data = {}, opts = {}) => {
    opts = Object.assign({}, {
      url: url,
      method: 'post',
      json: true,
      body: data,
    }, opts);
    return await ns.Request(opts);
  };
  ns.Request.get = async (url, qs = {}, opts = {}) => {
    opts = Object.assign({}, {
      url: url,
      method: 'get',
      json: true,
      qs: qs,
      headers: {
        'Cache-Control': 'no-cache'
      }
    }, opts);
    return await ns.Request(opts);
  };
  
  ns.ip2long = function ip2long(ip_address) {
    var output = false;
    if (ip_address.match(/^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$/)) {
      var parts = ip_address.split('.');
      var output = 0;
      output = ( parts [0] * Math.pow(256, 3) ) +
        ( parts [1] * Math.pow(256, 2) ) +
        ( parts [2] * Math.pow(256, 1) ) +
        ( parts [3] * Math.pow(256, 0) );
    }
    
    return output << 0;
  }
  
  ns.long2ip = function long2ip(proper_address) {
    proper_address = proper_address >>> 0;
    var output = false;  // www.
    
    if (!isNaN(proper_address) && ( proper_address >= 0 || proper_address <= 4294967295 )) {
      output = Math.floor(proper_address / Math.pow(256, 3)) + '.' +
        Math.floor(( proper_address % Math.pow(256, 3) ) / Math.pow(256, 2)) + '.' +
        Math.floor(( ( proper_address % Math.pow(256, 3) ) % Math.pow(256, 2) ) /
          Math.pow(256, 1)) + '.' +
        Math.floor(( ( ( proper_address % Math.pow(256, 3) ) % Math.pow(256, 2) ) %
          Math.pow(256, 1) ) / Math.pow(256, 0));
    }
    
    return output;
  };
  
  ns.toCamel = (letter) => {
    letter = letter.replace(/[-_ ](\w)/g, function (all, letter) {
      return letter.toUpperCase();
    });
    return letter;
  };
  
  ns.checksum = (data, include) => {
    let tempObj = MJSERVER._.pick(data, include);
    let keys = Object.keys(tempObj);
    keys.sort((a, b) => {
      return a.localeCompare(b);
    });
    let str = '';
    for (let i = 0; i < keys.length; i++) {
      str += tempObj[keys[i]];
    }
    return ns.md5(str);
  };
  
  ns.checkPrivateAPIAccess = (targetToken, rightToken) => {
    if (targetToken !== rightToken) {
      throw new MJSERVER.Exception(MJSERVER.errors.ERROR_TOKEN_CHECK);
    }
  };
  
  ns.rsa = (str, encoding = 'utf8') => {
    let hash = crypto.createHash('sha256');
    return hash.update(str, encoding).digest('base64');
  };
  
  ns.md5 = (str, encoding = 'utf8') => {
    return crypto.createHash('md5').update(str, encoding).digest('hex');
  };
  
  ns.encrypt = (str, type, encoding = 'utf8') => {
    return crypto.createHash(type).update(str, encoding).digest('hex');
  };
  
  ns.genSign = (data, excludes = [], keyValue, keyField = 'key', encoding = 'utf8') => {
    let keys = Object.keys(data);
    let useKeys = keys.filter(key => {
      return !~excludes.indexOf(key);
    });
    useKeys.sort((k1, k2) => {
      return k1.localeCompare(k2);
    });
    let paramsStrArr = [];
    for (let i = 0; i < useKeys.length; i++) {
      let key = useKeys[i];
      let val = data[key];
      if (val === null || val === undefined) {
        val = '';
      }
      paramsStrArr.push(`${key}=${val}`);
    }
    paramsStrArr.unshift(`version=1`);
    paramsStrArr.push(`${keyField}=${keyValue}`);
    
    let paramsStr = paramsStrArr.join('&');
    let sign = ns.md5(paramsStr, encoding);
    return sign;
  };
  ns.genHeePaySign = (data, key, seqNames = ['version', 'agent_id', 'agent_bill_id', 'agent_bill_time', 'pay_type', 'pay_amt', 'notify_url', 'user_ip',]) => {
    const orderNames = seqNames;
    let keys = Object.keys(data);
    let useKeys = keys.filter(key => {
      return ~orderNames.indexOf(key);
    });
    useKeys.sort((k1, k2) => {
      return orderNames.indexOf(k1) > orderNames.indexOf(k2) ? 1 : -1;
    });
    let paramsStrArr = [];
    for (let i = 0; i < useKeys.length; i++) {
      let key = useKeys[i];
      let val = data[key];
      if (val === null || val === undefined) {
        val = '';
      }
      paramsStrArr.push(`${key}=${val}`);
    }
    paramsStrArr.push(`key=${key}`);
    
    let paramsStr = paramsStrArr.join('&');
    let sign = ns.md5(paramsStr);
    data.sign = sign;
    return data;
  };
  
  ns.randStr = function (len, seed) {
    seed = seed || '0123456789';
    let result = [];
    for (let i = 0; i < len; i++) {
      result.push(seed[Math.floor(Math.random() * seed.length)]);
    }
    return result.join('');
  };
  
  ns.toBase64 = function (content, encoding = 'utf8') {
    return iconv.encode(content, encoding).toString('base64');
  }
  
  ns.fromBase64 = function (content, encoding = 'utf8') {
    return iconv.decode(new Buffer(content, 'base64'), encoding);
  };
  
  ns.getIPv4 = (ip, sep = '.') => {
    let userIp = ip || '127.0.0.1';
    userIp = userIp.replace(/[\s\S]*?(\d{1,3})\.(\d{1,3})\.(\d{1,3})\.(\d{1,3})[\s\S]*?/g, `$1${sep}$2${sep}$3${sep}$4`);
    return userIp;
  };
  
  ns.gb2312ToUtf8 = (str) => {
    let errMessage = '';
    if (!str) {
      return '';
    }
    try {
      let gb2312Str = str.replace(/%/g, ',0x').split(',');
      gb2312Str.shift();
      let buff = new Buffer(gb2312Str);
      errMessage = iconv.decode(buff, 'gb2312');
    }
    catch (e) {
      errMessage = '';
    }
    return errMessage;
  };
  ns.redisSetx = async function (key, value, ttl) {
    return new Promise((success, fail) => {
      MJSERVER.redis.setex(key, ttl, value, function (err, ret) {
        if (err) {
          return fail(err);
        }
        success(ret);
      });
    });
  };
  ns.redisGet = async function (key) {
    return new Promise((success, fail) => {
      MJSERVER.redis.get(key, function (err, ret) {
        if (err) {
          return fail(err);
        }
        success(ret);
      });
    });
  };
  
  ns.getAccessToken = async function getAccessToken() {
    
    const type = 'PUBLIC_ACCOUNT';
    var info = MJSERVER.config.WEICHAT_CONFIG[type];
    if (info == null) {
      throw new MJSERVER.Exception(-1, 'APP KEY Can\'t found.');
    }
    
    var opts = {
      url: MJSERVER.constants.OAUTH_INFO.getAccessToken2.url,
      method: 'get',
      json: false,
      qs: {
        appid: info.appid,
        secret: info.secret,
        grant_type: "client_credential"
      }
    };
    
    let res = await MJSERVER._.Request(opts);
    res = JSON.parse(res);
    if (res.errcode) {
      throw new MJSERVER.Exception(res.errcode, res.errmsg);
    }
    let ret = await ns.redisSetx(MJSERVER.constants.ACCESS_TOKEN_CACHE_KEY, res.access_token, 3600 * 2);
    console.log('refresh access token', Date.now());
    return res.access_token;
  };
  return ns;
}

module.exports = extendsFunc;